worker/upgradedatabase/worker.go

```diff
@@ -213,22 +212,26 @@ func (w *upgradeDB) upgradeDone() bool {
 	return false
 }
 
-func (w *upgradeDB) runUpgrade() {
+func (w *upgradeDB) runUpgrade() error {
 	w.logger.Infof("running database upgrade for %v on mongodb primary", w.toVersion)
 	w.setStatus(status.Started, fmt.Sprintf("upgrading database for %v", w.toVersion))
 
-	if err := w.agent.ChangeConfig(w.runUpgradeSteps); err == nil {
-		// Update the upgrade status document to unlock the other controllers.
-		if err = w.upgradeInfo.SetStatus(state.UpgradeDBComplete); err != nil {
-			w.logger.Errorf("failed to update upgrade info: %v", err)
-			w.setFailStatus()
-			return
-		}
-
-		w.logger.Infof("database upgrade for %v completed successfully.", w.toVersion)
-		w.setStatus(status.Started, fmt.Sprintf("database upgrade for %v completed", w.toVersion))
-		w.upgradeComplete.Unlock()
+	err := w.agent.ChangeConfig(w.runUpgradeSteps)
+	if err != nil {
+		w.setFailStatus()
+		return err
 	}
+	// Update the upgrade status document to unlock the other controllers.
+	err = w.upgradeInfo.SetStatus(state.UpgradeDBComplete)
+	if err != nil {
+		w.logger.Errorf("failed to update upgrade info: %v", err)
+		w.setFailStatus()
+		return err
```

Comment:
so we got all the way through an upgrade and fail setting the status to done, worker will bounce and start again. things should gracefully run through as no-ops a second time but it would be good to explicitly test this with a wrench or something to observe the behaviour

Reply:
Would you like tests/CI of this description to be included in this PR before merge?